src/mono/mono/metadata/marshal-lightweight.c

```diff
@@ -2424,22 +2430,29 @@ emit_unsafe_accessor_ctor_wrapper (MonoMethodBuilder *mb, MonoMethod *accessor_m
 	}
 
 	MonoType *target_type = sig->ret; // for constructors the return type is the target type
-	MonoClass *target_class = mono_class_from_mono_type_internal (target_type);
-
-	ERROR_DECL(find_method_error);
-	if (accessor_method->is_inflated) {
-		sig =  update_signature(accessor_method);
-		target_type = sig->ret;
-	}
 
 	if (target_type == NULL || m_type_is_byref (target_type) || unsafe_accessor_target_type_forbidden (target_type)) {
 		mono_mb_emit_exception_full (mb, "System", "BadImageFormatException", "Invalid usage of UnsafeAccessorAttribute.");
 		return;
 	}
+	
+	gboolean inflate_generic_data = FALSE;
+	if ((accessor_method->is_generic || to_be_inflated)) {
+		// We want to do a static lookup: target type must be G<T>, not just !T or !!T
+		g_assert (target_type->type != MONO_TYPE_VAR && target_type->type != MONO_TYPE_MVAR);
+		// If the target is generic, we'll find a generic MonoMethod* and then mark it to be
+		// inflated in order to handle instances, or gshared/gsharedvt instances.
+
+		inflate_generic_data = TRUE;
+	};
+
+	MonoClass *target_class = mono_class_from_mono_type_internal (target_type);
+
+	ERROR_DECL(find_method_error);
 
-	MonoMethodSignature *member_sig = ctor_sig_from_accessor_sig (mb, sig, ctx);
+	MonoMethodSignature *member_sig = ctor_sig_from_accessor_sig (mb, sig);
 	
-	MonoClass *in_class = mono_class_get_generic_type_definition (target_class);
+	MonoClass *in_class = target_class; // mono_class_get_generic_type_definition (target_class);
```

Comment:
Remove the comment here? It will be confusing to read later, since it will be the generic type already.